native: added shell support

aescolar opened the pull request:
* Now the native console driver also handles feeding stdin to the shell.
  So we can drive the shell from the command line, a pipe, or a file.
* Shell: added noprompt command to disable printing the prompt
* Added to the shell sample app and bluetooth shell tests the native port.
* Added a new config option for the POSIX arch (recommended stack size)
* An unrelated typo fix in a native_posix figure

aescolar committed:
native: added stdin handling for shell

Now the native console driver also handles stdin
so we can drive the shell from the command line,
a pipe or a file

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
test: shell: native_posix also supports shell

The native_posix port also supports shell apps thru its
stdin redirection.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
native: arch: added recommended stack size config

Added a new config variable with the recommended stack
size for threads which are only meant for the posix architecture

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

dbkinder commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -362,11 +369,51 @@ The following peripherals are currently provided with this board:
+
+  Commands will be feed to the shell as fast as the shell can process them.
+  To allow controlling the flow of commands to the shell, a special command has
+  been added ``!wait <ms>``.
```
Do you need to update the Shell documentation (in doc/subsystems/shell.rst ) to include a description for this command in the http://docs.zephyrproject.org/subsystems/shell.html#using-shell-commands section?  Perhaps this ``!wait`` is not actually a Shell command (since you mention it's handled by the console driver).

aescolar-ot replied:
As you mention, it is not really a shell command. I did not know how to call it other than "special command", as it really only affects the driver itself, and won't be seen by the shell.
So I think it may be a bit confusing to add it to the shell subsystem documentation.

dbkinder replied:
As you've clarified, I agree it shouldn't be documented as a shell command.  Maybe call it a "driver directive" or is that too odd?

aescolar-ot replied:
"Driver directive" is probably better. I will change it that way.
(But if later some better name pops to your mind just tell and I will update it)

aescolar-ot replied:
I rephrased a bit this paragraph. Please tell if it is clearer now, or if some changes would make it better. 
Like always, thanks you very much.

jhedberg commented on drivers/console/native_posix_console.c:
```diff
@@ -27,9 +39,199 @@ static int printk_init(struct device *arg)
+		FD_SET(STDIN_FILENO, &readfds);
+
+		ready = select(STDIN_FILENO+1, &readfds, NULL, NULL, &timeout);
+
```
I'd remove this empty line. Generally: no empty line between the assignment to a variable and testing for the variable's value.

jhedberg commented on drivers/console/native_posix_console.c:
```diff
@@ -27,9 +39,199 @@ static int printk_init(struct device *arg)
+		 * CONSOLE_MAX_LINE_LEN), or we reached EOF
+		 */
+		ret = fgets(cmd->line, CONSOLE_MAX_LINE_LEN, stdin);
+
```
I'd remove this empty line, since the test for the value is right after the assignment

jhedberg commented on drivers/console/native_posix_console.c:
```diff
@@ -27,9 +39,199 @@ static int printk_init(struct device *arg)
+		if (ret != 1) {
+			posix_print_error_and_exit("%s(): %s not understood, "
+					"syntax = !wait %%u\n",
+					__func__, s);
```
The alignment of the split lines doesn't look quite right (could be the GH user interface also distorting stuff)

aescolar-ot replied:
Should it be aligned with the parenthesis or 1 tab indented farther than the function call?

jhedberg commented on drivers/console/native_posix_console.c:
```diff
@@ -27,9 +39,199 @@ static int printk_init(struct device *arg)
+	}
+
+	posix_print_error_and_exit("%s(): '%s' not understood\n",
+				__func__, s);
```
Same here regarding the alignment

jhedberg commented on drivers/console/native_posix_console.c:
```diff
@@ -27,9 +39,199 @@ static int printk_init(struct device *arg)
+	struct timeval timeout;
+	fd_set readfds;
+	static struct console_input *cmd;
+	u32_t towait = CONFIG_NATIVE_STDIN_POLL_PERIOD;
```
Note that Zephyr timeouts are s32_t (e.g. the type that k_sleep expects) since K_FOREVER == -1. Is u32_t intentionally chosen since you don't need K_FOREVER?

jhedberg reviewed (changes requested):
Functionally the PR looks quite good to me. Just some minor coding-style related comments.

dbkinder reviewed (approved):
works for me, thanks!

aescolar-ot commented on drivers/console/native_posix_console.c:
```diff
@@ -27,9 +39,199 @@ static int printk_init(struct device *arg)
+		}
+
+		/* Remove a possible end of line */
+		int last = (int)strlen(cmd->line) - 1;
```
Regarding the cast. No, I guess it is not really needed as we target x86. I was being (unnecessarily) portability pedantic, for the case in which length is 0, and the sizeof(size_t) != size(int). Thinking about it (if I'm not mistaken), I don't think there would be a problem with smaller types, as 1 is int, and would promotes the whole operation to int.
In the case of sizeof(size_t) > sizeof(int), the whole operation would promote to size_t's size as unsigned, wrap to that unsigned type MAX -1, and then when casted down to int be implementation specific(?). [=> felt safer and faster to just add the cast]

Would you prefer me to remove it?

aescolar committed:
shell: Added noprompt command

Added noprompt command to shell. It will disable printing
the prompt.
For the native port, when feeding commands from a file or
pipe the prompt reprinting after each command (without echoing)
just confuses the user.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
native: doc: typo in figure

Corrected a typo in native_posix's architecture figure

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar-ot commented:
A few changes:
* Bugfix: when feeding from a pipe the driver was actually not stalling the process
* Added !quit driver directive
* Some minor refactoring of the code
* Shell: added noprompt command to disable the prompt (the prompt is quite anoying when feeding from a file/pipe).
* (freebie) An unrelated typo fix

@dbkinder : The documentation changes since your last review are limited to the new "driver directives" subsection, a change in the ``Commands will be feed to the..`` paragraph, and the changes in shell.rst
@jhedberg : All your feedback should be accounted for now (except for the int casting)

pfalcon commented on arch/posix/Kconfig:
```diff
@@ -25,6 +25,16 @@ config ARCH_DEFCONFIG
+config ARCH_POSIX_RECOMMENDED_STACK_SIZE
+	int
+	depends on ARCH_POSIX
+	default 24
```
It's weird to be told about "stack size" and be shown "24, in bytes". Something got to be wrong here (e.g., it's not really *stack size* per se).

aescolar-ot replied:
It is the stack you allocate for a thread like for example here
https://github.com/aescolar-ot/zephyr/blob/01d0470a08a2288403932300ea072551284467d4/drivers/console/native_posix_console.c#L66

In the POSIX arch that stack is actually not used to hold the thread stack. The real stack of the thread is a normal pthread native stack.
That's why it can be so small.

pfalcon replied:
So, maybe call it DUMMY_STACK, or something. That maybe long, but at least avoid confusion (or otherwise it won't be soon when other people will understand the native port well enough ;-) ).

aescolar-ot replied:
That parameter will only be used for drives (or similar) meant only for the POSIX arch. So it is a bit of a hidden configuration option. I can expand the help description to clarify it though.

pfalcon commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -361,11 +368,66 @@ The following peripherals are currently provided with this board:
+forwarded to the shell.
+These directives are:
+
+- ``!wait <ms>``: When received, the driver will pause feeding commands to the
```
I would consider such things "dangerous".  User data is user data - why trick with it?

aescolar-ot replied:
These are "inband commands" meant to affect only the transport driver. They start with `!` so you can tell them apart from normal shell commands, and so that they will not conflict with future normal shell commands.
For the use case they are meant for (You have a file with all commands you want to send), it makes most sense to have them inline. For ex, a file with a content like:
```
noprompt
select kernel
uptime
!wait 500
uptime
!quit
```
(Imagine those commands are something more sensible, like for BLE commands)

aescolar-ot replied:
>for a good way to #ifdef all of that cod
All that code is ifdef'ed under NATIVE_POSIX_STDIN_CONSOLE

pfalcon replied:
Yeah, it's just not everyone knows that besides this RTOSesquely backward "console" which you implement here, Zephyr also has standard POSIXy console with getchar()/putchar() API. My concern was that implementation of that would conflict with the changes here (because they both call getchar()), but somehow it seems to work so far: https://github.com/zephyrproject-rtos/zephyr/pull/5947
